diff --git a/init.lua b/init.lua
index 6c8abd7..322a21d 100644
--- a/init.lua
+++ b/init.lua
@@ -4,6 +4,7 @@ cudnn = require 'cudnn.env'
 require('cudnn.ffi')
 local C = cudnn.C
 local ffi = require 'ffi'
+local THC = ffi.os == 'Windows' and ffi.load('THC') or ffi.C
 
 --------------------------------------------------------------------
 -- defaults, each should be overrideable via env var:
@@ -151,7 +152,7 @@ end
 
 function cudnn.call(f, ...)
     C.cudnnSetStream(cudnn.getHandle(),
-                     ffi.C.THCState_getCurrentStream(cutorch.getState()))
+                     THC.THCState_getCurrentStream(cutorch.getState()))
     return C[f](...)
 end
 
