diff --git a/lib/THC/THCThrustAllocator.cuh b/lib/THC/THCThrustAllocator.cuh
index 3ab2d37..5e8a6c5 100644
--- a/lib/THC/THCThrustAllocator.cuh
+++ b/lib/THC/THCThrustAllocator.cuh
@@ -1,6 +1,8 @@
 #ifndef THC_THRUST_ALLOCATOR_INC
 #define THC_THRUST_ALLOCATOR_INC
 
+#include <cstddef>
+
 /// Allocator for Thrust to re-route its internal device allocations
 /// to the THC allocator
 class THCThrustAllocator {
