diff --git a/fpconv.h b/fpconv.h
index 7b0d0ee..f674e41 100644
--- a/fpconv.h
+++ b/fpconv.h
@@ -4,6 +4,10 @@
  *
  * Longest double printed with %.14g is 21 characters long:
  * -1.7976931348623e+308 */
+#ifdef _MSC_VER
+#define inline __inline
+#define snprintf _snprintf
+#endif
 # define FPCONV_G_FMT_BUFSIZE   32
 
 #ifdef USE_INTERNAL_FPCONV
diff --git a/lua-cjson-2.1devel-1.rockspec b/lua-cjson-2.1devel-1.rockspec
index 154e333..0f7652e 100644
--- a/lua-cjson-2.1devel-1.rockspec
+++ b/lua-cjson-2.1devel-1.rockspec
@@ -47,6 +47,7 @@ build = {
     -- Override default build options (per platform)
     platforms = {
         win32 = { modules = { cjson = { defines = {
+            "strncasecmp=_strnicmp",
             "DISABLE_INVALID_NUMBERS"
         } } } }
     },
diff --git a/strbuf.c b/strbuf.c
index ac779e4..c3272f0 100644
--- a/strbuf.c
+++ b/strbuf.c
@@ -94,8 +94,8 @@ void strbuf_set_increment(strbuf_t *s, int increment)
 static inline void debug_stats(strbuf_t *s)
 {
     if (s->debug) {
-        fprintf(stderr, "strbuf(%lx) reallocs: %d, length: %d, size: %d\n",
-                (long)s, s->reallocs, s->length, s->size);
+        fprintf(stderr, "strbuf(%p) reallocs: %d, length: %d, size: %d\n",
+                s, s->reallocs, s->length, s->size);
     }
 }
 
@@ -168,8 +168,8 @@ void strbuf_resize(strbuf_t *s, int len)
     newsize = calculate_new_size(s, len);
 
     if (s->debug > 1) {
-        fprintf(stderr, "strbuf(%lx) resize: %d => %d\n",
-                (long)s, s->size, newsize);
+        fprintf(stderr, "strbuf(%p) resize: %d => %d\n",
+                s, s->size, newsize);
     }
 
     s->size = newsize;
diff --git a/strbuf.h b/strbuf.h
index d861108..b135560 100644
--- a/strbuf.h
+++ b/strbuf.h
@@ -24,6 +24,9 @@
 
 #include <stdlib.h>
 #include <stdarg.h>
+#ifdef _MSC_VER
+#define inline __inline
+#endif
 
 /* Size: Total bytes allocated to *buf
  * Length: String length, excluding optional NULL terminator.
