diff --git a/init.lua b/init.lua
index 167843d..67ab5e4 100644
--- a/init.lua
+++ b/init.lua
@@ -46,10 +46,15 @@ local function isWindows()
       package.config:sub(1,1) == '\\'
 end
 
-if isWindows() or (not cutils.isatty()) then
+local hasTPut = true -- default true for non windows
+if isWindows() then
+  hasTPut = sys.fexecute('where tput'):find('tput')
+end
+
+if not hasTPut or not cutils.isatty() then
    noColors()
 else
-   local outp = os.execute('tput colors >/dev/null')
+   local outp = os.execute('tput colors >' .. (isWindows() and 'NUL' or '/dev/null'))
    if type(outp) == 'boolean' and not outp then
       noColors()
    elseif type(outp) == 'number' and outp ~= 0 then
@@ -520,6 +525,10 @@ local aliases = [[
    alias lla='ls -lahF';
 ]]
 
+if isWindows() then
+   aliases = ""
+end
+
 -- Penlight
 pcall(require,'pl')
 
diff --git a/readline.c b/readline.c
index e14c867..42b8c8a 100644
--- a/readline.c
+++ b/readline.c
@@ -4,8 +4,12 @@
 #include "lua.h"
 #include "lauxlib.h"
 #include "lualib.h"
+#ifndef WinEditLine
 #include <readline/readline.h>
 #include <readline/history.h>
+#else
+#include <editline/readline.h>
+#endif
 #include <ctype.h>
 
 #if LUA_VERSION_NUM == 501
@@ -260,12 +264,16 @@ static int f_setup(lua_State *L)
   /* Break words at every non-identifier character except '.' and ':'. */
   rl_completer_word_break_characters =
     "\t\r\n !\"#$%&'()*+,-/;<=>?@[\\]^`{|}~";
+#ifndef WinEditLine
   rl_completer_quote_characters = "\"'";
+#endif
 /* #if RL_READLINE_VERSION < 0x0600 */
   rl_completion_append_character = '\0';
 /* #endif */
   rl_attempted_completion_function = lua_rl_complete;
+#ifndef WinEditLine
   rl_initialize();
+#endif
 
   return 0;
 }
diff --git a/trepl-scm-1.rockspec b/trepl-scm-1.rockspec
index 54269b0..56d9627 100644
--- a/trepl-scm-1.rockspec
+++ b/trepl-scm-1.rockspec
@@ -40,10 +40,10 @@ build = {
 		    ['readline'] = {
                sources = {'readline.c'},
                libraries = {'readline'},
-			   defines = {"USE_READLINE_STATIC"},
-               incdirs = {"windows"},
-               libdirs = {"windows"},
-               libraries = {'readline-win'}
+               defines = {"WinEditLine"},
+               incdirs = {"..\\..\\win-files\\3rd\\wineditline-2.101\\include"},
+               libdirs = {"..\\..\\win-files\\3rd\\wineditline-2.101\\lib64"},
+               libraries = {'edit_static', 'user32'}
 			}
 		 }
 	  }
diff --git a/utils.c b/utils.c
index 62a3e4e..b446051 100644
--- a/utils.c
+++ b/utils.c
@@ -13,7 +13,7 @@
 
 int treplutils_isatty(lua_State *L)
 {
-  lua_pushboolean(L, 0);
+  lua_pushboolean(L, _isatty(1));
   return 1;
 }
 
